docs(tip-1063): clarify feature registry, scheduling and activation (#4114)
tempoxyz/tempo: tips/tip-1063.md

## Overview

A feature flag controls whether a specific Tempo protocol change is active. Feature gated code ships inactive in a release. Each feature is recorded in an onchain feature registry with a minimum supported Tempo version. A validator supports the feature when its reported version is greater than or equal to that `minimum_supported_version`.

> A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the feature is scheduled, Tempo validates that enough active validators support the feature, and the activiation timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.

Reviewer: ```suggestion
A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the feature is scheduled, Tempo validates that enough active validators support the feature, and the activation timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.
```

## Feature Registry

Tempo stores protocol feature data in an onchain feature registry. For a given feature, the registry records the feature ID, minimum supported version, status, and activation timestamp. Registry state is the source of truth for feature activation on Tempo.

> The registry specifies a designated signer authorized to add new features and schedule activation timestamps for registered features.

Reviewer: ```suggestion
The registry specifies a designated signer authorized to register approved features onchain and schedule activation timestamps.
```

Each registry entry includes:

- `feature_id`: a globally unique, lowercase, dot-separated feature identifier

> - `description`: a short description of the behavior; linking the TIP

Reviewer: ```suggestion
- `description`: a short description of the behavior; including the TIP reference
```

- `minimum_supported_version`: the lowest Tempo version expected to validate, build, simulate, and serve the feature correctly after activation

> - `status`: the status of the feature. Available statuses are `pending`, `active`, and `cancelled`

Reviewer: ```suggestion
- `status`: the status of the feature. Available statuses are `pending` and `active`
```

- `activation_timestamp`: the timestamp when the feature becomes active, if scheduled; set to 0 means the feature is not scheduled

```solidity
contract FeatureRegistry {
  // --snip--

  struct Feature {
    string description;
    string minimum_supported_version;
    Status status;
    uint64 activation_timestamp;
  }

  mapping(string feature_id => Feature) features;

  // --snip--
}
```

Registered feature identity is immutable. After registration, the registry must not allow edits to `feature_id` or `minimum_supported_version`.

Feature lifecycle state is mutable. Mutable state is limited to `status` and `activation_timestamp`.

The registry also tracks the Tempo version each validator reports. This is used to ensure that enough active validators are running a Tempo version that supports a given feature before the feature can be scheduled for activation. A version report counts only if it is deterministic from chain history. For a given feature, a validator supports the feature when its latest reported Tempo version is greater than or equal to the feature's `minimum_supported_version`.

> Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and are not part of this design.

Reviewer: ```suggestion
Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and is not part of this design for simplicity.
```

A feature has quorum when enough validators in the active validator set support it. The threshold is Tempo's consensus threshold, expressed as `n/m`, where `m` is the active validator set size and `n` is the required number of active validators that support the feature. Quorum does not activate the feature by itself. It only makes the feature eligible to be scheduled for activation.

> The registry should expose enough read methods for nodes, indexers, and operators to inspect feature state, validator reported versions, current support counts, and the global activation quorum threshold. Callers should be able to determine whether a feature is known, scheduled, active, or cancelled, including for a specific block or timestamp.

Reviewer: ```suggestion
The registry should expose enough read methods for nodes, indexers, and operators to inspect feature state, validator reported versions, current support counts, and the global activation quorum threshold. Callers should be able to determine whether a feature is known, scheduled, or active, including for a specific block or timestamp.
```

Registry state changes must be observable through events for registration, activation scheduling, schedule updates, support count changes, activation, and cancellation.

## Scheduling and Activation

>  registered feature can be scheduled for activation once the active validator set has reached quorum for the feature's `minimum_supported_version`. The authorised signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and quorum over active validators is satisfied. The requested activation timestamp must also be in the future.

Reviewer: ```suggestion
A registered feature can be scheduled for activation once the active validator set has reached quorum for the feature's `minimum_supported_version`. The authorised signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and quorum over active validators is satisfied. The requested activation timestamp must also be in the future.
```

> A registered feature can be scheduled for activation once the active validator set has reached quorum over the minimum supported version that the network is running. The authorized signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and quorum over active validators is satisfied. The requested activation timestamp must also be in the future.

Reviewer: ```suggestion
 registered feature can be scheduled for activation once the active validator set has reached quorum for the feature's `minimum_supported_version`. The authorised signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and quorum over active validators is satisfied. The requested activation timestamp must also be in the future.
```

A scheduled feature becomes active when block time reaches its activation timestamp and activation is applied during the pre-execution hooks at the start of block processing. If a feature's activation timestamp has been reached, the registry marks the feature `active` and emits the activation event. User transactions in that block are processed with the feature active.

Nodes derive the active feature set from registry state. Validators, RPC nodes, and other infrastructure must use the same active feature state when evaluating protocol behavior for a block. Once a feature is active, any node that validates, builds, simulates, or serves protocol behavior at or after the activation timestamp must support that feature.

## Cancellation and Supersession

Before activation, the authorised signer can cancel a feature's scheduled activation. Cancellation clears `activation_timestamp` and leaves the feature `pending`. The feature can be scheduled again later.

> Before activation, the authorized signer can also reschedule a pending feature by replacing its activation timestamp with a new future timestamp.

Reviewer: ```suggestion
Before activation, the authorized signer can cancel a feature's scheduled activation. Cancellation clears `activation_timestamp` and leaves the feature `pending`. The feature can be scheduled again later.
```

> Before activation, the authorized signer can cancel a feature's scheduled activation. Cancellation clears `activation_timestamp` and leaves the feature `pending`. The feature can be scheduled again later.

Reviewer: ```suggestion
```

    enum FeatureStatus {
        Pending,
        Active,
        Cancelled
    }

    struct Feature {
        string description;
        string minimumSupportedVersion;
        FeatureStatus status;
        uint64 activationTimestamp;
    }

    /// @notice Returns the address authorized to register features and schedule activation.
    function authorizedSigner() external view returns (address);

    /// @notice Returns the global validator quorum threshold.
    function activationQuorum() external view returns (uint256 numerator, uint256 denominator);

    /// @notice Registers a new pending feature.
    function registerFeature(
        string calldata featureId,
        string calldata description,
        string calldata minimumSupportedVersion
    ) external;

    /// @notice Schedules activation for a pending feature.
    function scheduleActivation(string calldata featureId, uint64 activationTimestamp) external;

    /// @notice Replaces the activation timestamp for a scheduled pending feature.
    function rescheduleActivation(string calldata featureId, uint64 activationTimestamp) external;

    /// @notice Cancels a pending feature before activation.

>     function cancelFeature(string calldata featureId) external;

Reviewer: ```suggestion
    /// @notice Cancels a scheduled activation before the feature become active.
    function cancelScheduledActivation(string calldata featureId) external;
```

    function validatorVersion(address validator) external view returns (string memory);

    /// @notice Returns whether a validator's reported version supports a feature.
    function validatorSupportsFeature(address validator, string calldata featureId) external view returns (bool);

    /// @notice Returns current active-validator support for a feature.
    function featureSupportCount(string calldata featureId) external view returns (uint256 support, uint256 required);

    /// @notice Returns whether a feature has quorum support from the active validator set.
    function hasQuorum(string calldata featureId) external view returns (bool);
}
```

### Events

```solidity
/// @notice Emitted when a feature is registered.
event FeatureRegistered(string featureId, string minimumSupportedVersion);

/// @notice Emitted when a pending feature is scheduled for activation.
event FeatureScheduled(string featureId, uint64 activationTimestamp);

/// @notice Emitted when a pending feature's activation timestamp is replaced.
event FeatureRescheduled(string featureId, uint64 activationTimestamp);

/// @notice Emitted when a pending feature is cancelled.

> event FeatureCancelled(string featureId);

Reviewer: ```suggestion
/// @notice Emitted when a scheduled feature is cancelled.
event FeatureScheduleCancelled(string featureId);
```

> A feature flag controls whether a specific Tempo protocol change is active. Feature gated code ships inactive in a release. Each feature is recorded in an onchain feature registry with a minimum supported Tempo version. A validator supports the feature when its reported version is greater than or equal to that minimum specified version.

Reviewer: ```suggestion
A feature flag controls whether a specific Tempo protocol change is active. Feature gated code ships inactive in a release. Each feature is recorded in an onchain feature registry with a minimum supported Tempo version. A validator supports the feature when its reported version is greater than or equal to that `minimum_supported_version`.
```

A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the feature is scheduled, Tempo validates that enough active validators support the feature, and the activiation timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.

## Feature Registry

Tempo stores protocol feature data in an onchain feature registry. For a given feature, the registry records the feature ID, minimum supported version, status, and activation timestamp. Registry state is the source of truth for feature activation on Tempo.

The registry specifies a designated signer authorized to add new features and schedule activation timestamps for registered features.

Each registry entry includes:

- `feature_id`: a globally unique, lowercase, dot-separated feature identifier

> - `description`: a short description of the behavior, incl the TIP

Reviewer: ```suggestion
- `description`: a short description of the behavior; linking the TIP
```

Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and are not part of this design.

A feature has quorum when enough validators in the active validator set support it. The threshold is Tempo's consensus threshold, expressed as `n/m`, where `m` is the active validator set size and `n` is the required number of active validators that support the feature. Quorum does not activate the feature by itself. It only makes the feature eligible to be scheduled for activation.

The registry should expose enough read methods for nodes, indexers, and operators to inspect feature state, validator reported versions, current support counts, and the global activation quorum threshold. Callers should be able to determine whether a feature is known, scheduled, active, or cancelled, including for a specific block or timestamp.

Registry state changes must be observable through events for registration, activation scheduling, schedule updates, support count changes, activation, and cancellation.

## Scheduling and Activation

> A registered feature can be scheduled for activation once the active validator set has reached quorum over the minimum supported version that the network is running. The authorized signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and the required version for a given feature is <= the network's minimum active version. The requested activation timestamp must also be in the future.

Reviewer: ```suggestion
A registered feature can be scheduled for activation once the active validator set has reached quorum over the minimum supported version that the network is running. The authorized signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and quorum over active validators is satisfied. The requested activation timestamp must also be in the future.
```

A feature flag controls whether a specific Tempo protocol change is active. Feature gated code ships inactive in a release. Each feature is recorded in an onchain feature registry with a minimum supported Tempo version. A validator supports the feature when its reported version is greater than or equal to that minimum specified version.

> A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the activation timestamp is scheduled, Tempo validates that enough active validators support the feature, and that timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.

Reviewer: ```suggestion
A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the feature is scheduled, Tempo validates that enough active validators support the feature, and the activiation timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.
```

## Feature Registry

Tempo stores protocol feature data in an onchain feature registry. For a given feature, the registry records the feature ID, minimum supported version, status, and activation timestamp. Registry state is the source of truth for feature activation on Tempo.

The registry specifies a designated signer authorized to add new features and schedule activation timestamps for registered features.

Each registry entry includes:

- `feature_id`: a globally unique, lowercase, dot-separated feature identifier
- `description`: a short description of the behavior, incl the TIP
- `minimum_supported_version`: the lowest Tempo version expected to validate, build, simulate, and serve the feature correctly after activation
- `status`: the status of the feature. Available statuses are `pending`, `active`, and `cancelled`

> - `activation_timestamp`: the timestamp when the feature becomes active, if scheduled

Reviewer: ```suggestion
- `activation_timestamp`: the timestamp when the feature becomes active, if scheduled; set to 0 means the feature is not scheduled
```

Scheduling writes the activation timestamp to the feature metadata, an event is emitted to notify network participants and the scheduled timestamp becomes the upgrade deadline for any nodes running a version older than the minimum supported version. Network quorum on version support is validated when the timestamp is scheduled and is not revalidated at activation time.

A scheduled feature becomes active when block time reaches its activation timestamp and activation is applied during the pre-execution hooks at the start of block processing. If a feature's activation timestamp has been reached, the registry marks the feature `active` and emits the activation event. User transactions in that block are processed with the feature active.

Nodes derive the active feature set from registry state. Validators, RPC nodes, and other infrastructure must use the same active feature state when evaluating protocol behavior for a block. Once a feature is active, any node that validates, builds, simulates, or serves protocol behavior at or after the activation timestamp must support that feature.

## Cancellation and Supersession

> A feature can be cancelled after registration and before activation. This includes features that have reached quorum, and features that already have a scheduled activation timestamp.

Reviewer: ```suggestion
Before activation, the authorised signer can cancel a feature's scheduled activation. Cancellation clears `activation_timestamp` and leaves the feature `pending`. The feature can be scheduled again later.
```

> Cancelling a feature marks its status as `cancelled` in the feature registry. If the feature had a scheduled activation timestamp, cancellation clears the timestamp. A cancelled feature can be rescheduled for activation at a later date under the same `feature_id`.

Reviewer: ```suggestion
```

> A feature can be scheduled for activation once enough active validators have signaled support. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the activation timestamp is scheduled, Tempo validates that enough active validators support the feature, and that timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.

Reviewer: ```suggestion
A feature can be scheduled for activation once enough active validators have reported a Tempo version that supports it. A designated multisig schedules the feature by setting a future activation timestamp in an onchain feature registry. When the activation timestamp is scheduled, Tempo validates that enough active validators support the feature, and that timestamp becomes the upgrade deadline. When block time reaches the timestamp, the feature becomes active. Validators that have not upgraded must stop before processing blocks that require the active feature.
```

## Feature Registry

Tempo stores protocol feature data in an onchain feature registry. For a given feature, the registry records the feature ID, minimum supported version, status, and activation timestamp. Registry state is the source of truth for feature activation on Tempo.

The registry specifies a designated signer authorized to add new features and schedule activation timestamps for registered features.

Each registry entry includes:

- `feature_id`: a globally unique, lowercase, dot-separated feature identifier

> - `description`: a short description of the behavior

Reviewer: ```suggestion
- `description`: a short description of the behavior, incl the TIP
```

- `minimum_supported_version`: the lowest Tempo version expected to validate, build, simulate, and serve the feature correctly after activation
- `status`: the status of the feature. Available statuses are `pending`, `active`, and `cancelled`

> - `activation_timestamp`: the timestamp when the feature becomes active, if scheduled

Reviewer 2: and if not its just active immediately or with delay?

Reviewer: if it is not scheduled, it does not activate. `activation_timestamp = 0` means unscheduled. scheduling sets it to a future timestamp, and the feature activates when that timestamp is reached

```solidity
contract FeatureRegistry {
  // --snip--

  struct Feature {
    string description;
    string minimum_supported_version;
    Status status;
    uint64 activation_timestamp;
  }

  mapping(string feature_id => Feature) features;

  // --snip--
}
```

> Registered feature metadata is immutable. After registration, the registry must not allow edits to feature fields, including `feature_id` or `minimum_supported_version`. Mutable registry state is limited to status and activation timestamp.

Reviewer: ```suggestion
Registered feature identity is immutable. After registration, the registry must not allow edits to `feature_id` or `minimum_supported_version`.

Feature lifecycle state is mutable. Mutable state is limited to `status` and `activation_timestamp`.
```

> The registry also tracks the Tempo version each validator reports. This is used to ensure that enough active validators are running a Tempo version that supports a given feature before the feature can be scheduled for activation. A version report counts only if it is deterministic from chain history. For a given feature, a validator supports the feature when its latest reported Tempo version is greater than or equal to the feature's `minimum_supported_version`.

Reviewer 2: might need to double click on this point: who updates the registry?

Reviewer: agreed, we should make this explicit. i’d say a validator’s reported version can only be updated by that validator through some protocol-defined reporting path, while the multisig updates feature lifecycle state

> Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and are not part of this design.

Reviewer 2: ```suggestion
Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and are not part of this design for simplicity. 
```

Releases are monotonic for feature support. A newer release supports every feature supported by earlier releases. Selectively opting out of features or non-monotonic support would require per-feature signaling and are not part of this design.

A feature has quorum when enough validators in the active validator set support it. The threshold is Tempo's consensus threshold, expressed as `n/m`, where `m` is the active validator set size and `n` is the required number of active validators that support the feature. Quorum does not activate the feature by itself. It only makes the feature eligible to be scheduled for activation.

The registry should expose enough read methods for nodes, indexers, and operators to inspect feature state, validator reported versions, current support counts, and the global activation quorum threshold. Callers should be able to determine whether a feature is known, scheduled, active, or cancelled, including for a specific block or timestamp.

Registry state changes must be observable through events for registration, activation scheduling, schedule updates, support count changes, activation, and cancellation.

## Scheduling and Activation

> A registered feature can be scheduled for activation once the active validator set has reached quorum over the minimum supported version that the network is running. The authorized signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is `pending` and the required version for a given feature is <= the network's minimum active version. The requested activation timestamp must also be in the future.

Reviewer 2: ```suggestion
A registered feature can be scheduled for activation once the active validator set has reached quorum over the minimum supported version that the network is running. The authorized signer/multisig specified by the registry contract then schedules activation by submitting a registry transaction with the feature ID and a future activation timestamp. The registry accepts the scheduling transaction only if the feature is registered, the status is pending, the requested activation timestamp is in the future, and the feature has quorum support from the active validator set. A feature has quorum support when at least the required number of active validators have reported a Tempo version greater than or equal to the feature’s minimum_supported_version.
```

A scheduled feature becomes active when block time reaches its activation timestamp and activation is applied during the pre-execution hooks at the start of block processing. If a feature's activation timestamp has been reached, the registry marks the feature `active` and emits the activation event. User transactions in that block are processed with the feature active.

Nodes derive the active feature set from registry state. Validators, RPC nodes, and other infrastructure must use the same active feature state when evaluating protocol behavior for a block. Once a feature is active, any node that validates, builds, simulates, or serves protocol behavior at or after the activation timestamp must support that feature.

## Cancellation and Supersession

A feature can be cancelled after registration and before activation. This includes features that have reached quorum, and features that already have a scheduled activation timestamp.

> Cancelling a feature marks its status as `cancelled` in the feature registry. If the feature had a scheduled activation timestamp, cancellation clears the timestamp. A cancelled feature can be rescheduled for activation at a later date under the same `feature_id`.

Reviewer 2: ```suggestion
Cancelling a feature marks its status as `cancelled` in the feature registry. If the feature had a scheduled activation timestamp, cancellation clears the timestamp. A cancelled feature can be scheduled again under the same feature_id. Scheduling a cancelled feature transitions its status back to pending and rechecks quorum at scheduling time.
```

Before activation, the authorized signer can also reschedule a pending feature by replacing its activation timestamp with a new future timestamp.

After activation, a feature cannot be cancelled or locally disabled. Active features are part of chain history, and removing or toggling off their code can break syncing, state assumptions, or protocol behavior where the feature was active.

If an active feature needs to be fixed, disabled, or replaced, the change ships as a new protocol feature with its own `feature_id` and `minimum_supported_version`. For example, a fix to `fee_amm` can be registered as `fee_amm_rounding_fix`. The original `fee_amm` feature remains available for historical execution, and the new feature gates the corrected behavior from its own activation point.

## Interfaces

### Feature Registry

The feature registry exposes the following interface:

```solidity
// SPDX-License-Identifier: MIT

> pragma solidity ^0.8.13;

Reviewer: ```suggestion
```

    function activationQuorum() external view returns (uint256 numerator, uint256 denominator);

    /// @notice Registers a new pending feature.
    function registerFeature(
        string calldata featureId,
        string calldata description,
        string calldata minimumSupportedVersion
    ) external;

    /// @notice Schedules activation for a pending feature.
    function scheduleActivation(string calldata featureId, uint64 activationTimestamp) external;

    /// @notice Replaces the activation timestamp for a scheduled pending feature.
    function rescheduleActivation(string calldata featureId, uint64 activationTimestamp) external;

    /// @notice Cancels a pending feature before activation.
    function cancelFeature(string calldata featureId) external;

    /// @notice Returns the registered feature metadata and current status.
    function getFeature(string calldata featureId) external view returns (Feature memory);

    /// @notice Returns whether a feature is registered.
    function isFeatureRegistered(string calldata featureId) external view returns (bool);

    /// @notice Returns whether a feature is active at the current block.

>     function isFeatureActive(string calldata featureId) external view returns (bool);

Reviewer 2: would also add to check when a feature become active 

```
/// @notice Returns whether a feature is active at a given timestamp.
function isFeatureActiveAt(string calldata featureId, uint64 timestamp) external view returns (bool);
```

Reviewer: “active at timestamp” can be ambiguous because chain history is block-based. two blocks can have timestamps around boundaries, reorgs exist, and the registry state itself changes over blocks
